src/upkeeps/UpKeepManager.sol

```diff
@@ -0,0 +1,497 @@
+    // PUBLIC: Governance
+    ////////////////////////////////////////////////////////////////////////////
+
+    /// @dev It will initiate the upKeep job for monitoring members
+    /// @notice only callable via governance
+    /// @param _gasLimit gas limit for the upKeepManager monitoring upkeep task
+    function initializeBaseUpkeep(uint256 _gasLimit) external onlyGovernance {
+        if (_gasLimit == 0) revert ZeroUintValue();
+
+        monitoringUpKeepId = _registerUpKeep(address(this), _gasLimit, NAME);
+
+        if (monitoringUpKeepId > 0) {
+            /// @dev give allowance of spending LINK funds
+            LINK.approve(address(CL_REGISTRY), type(uint256).max);
+        }
+    }
+
+    /// @dev Adds an member into the manager
+    /// @notice only callable via governance
+    /// @param _memberAddress contract address to be register as new member
+    /// @param _name member's name
+    /// @param _gasLimit amount of gas to provide the target contract when performing upkeep
+    /// @param _existingUpKeepId optional upKeepId to add member which has being registered from other admins
+    function addMember(address _memberAddress, string memory _name, uint256 _gasLimit, uint256 _existingUpKeepId)
+        external
+        onlyGovernance
+    {
+        /// @dev sanity checks before adding a new member in storage
+        if (_memberAddress == address(0)) revert ZeroAddress();
+        if (membersInfo[_memberAddress].gasLimit != 0) revert MemberAlreadyRegister(_memberAddress);
+        if (_gasLimit == 0) revert ZeroUintValue();
+        if (bytes(_name).length == 0) revert EmptyString();
+
+        _members.add(_memberAddress);
+
+        // NOTE: when `_existingUpKeepId` is greater than zero, assumes that other admin has registered it
+        //       and there are not needs to register again
+        uint256 upKeepId = _existingUpKeepId > 0 ? _existingUpKeepId : _registerUpKeep(_memberAddress, _gasLimit, _name);
+
+        membersInfo[_memberAddress] = MemberInfo({name: _name, gasLimit: _gasLimit, upKeepId: upKeepId});
+
+        emit NewMember(_memberAddress, _name, _gasLimit, upKeepId, block.timestamp);
+    }
+
+    /// @dev Cancels an member's upkeep job
+    /// @notice only callable via governance
+    /// @param _memberAddress contract address to be cancel upkeep
+    function cancelMemberUpKeep(address _memberAddress) external onlyGovernance {
+        if (!_members.contains(_memberAddress)) revert NotMemberIncluded(_memberAddress);
+
+        // NOTE: only member which upkeep is being cancelled can be removed
+        uint256 upKeepId = membersInfo[_memberAddress].upKeepId;
+        CL_REGISTRY.cancelUpkeep(upKeepId);
+    }
+
+    /// @dev Withdraws LINK funds and remove member from manager
+    /// @notice only callable via governance
+    /// @param _memberAddress contract address to be remove from manager
+    function withdrawLinkFundsAndRemoveMember(address _memberAddress) external onlyGovernance {
```

Comment:
did you consider including this in the check/perform upkeep as well? the withdrawal could be initiated by a keeper imo, once the 50 blocks have been reached. would save user (techops) having to make a second call >50 blocks later.

you are already looping through the members in the check upkeep, maybe in that same loop you could check if the member is in the process of being cancelled, and then call `withdrawLinkFundsAndRemoveMember` for it if 50 blocks passed?